utils/contract-utils/src/contract_context.rs

```diff
@@ -0,0 +1,34 @@
+use casper_contract::unwrap_or_revert::UnwrapOrRevert;
+use casper_types::{system::CallStackElement, Key};
+
+use crate::ContractStorage;
+
+pub trait ContractContext<Storage: ContractStorage> {
+    fn storage(&self) -> &Storage;
+
+    fn get_caller(&self) -> Key {
+        let call_stack = self.storage().call_stack();
+        let caller = call_stack.get(call_stack.len() - 2);
+        element_to_key(caller.unwrap_or_revert())
+    }
```

Comment:
To make sure I understand this line, the purpose of using `get_caller` as defined here is to let contracts own tokens in case a contract called a token's entrypoint. If so, I'd recommend renaming it not to confuse this with `runtime::get_caller.`

```suggestion
    fn get_immediate_caller(&self) -> Key {
        let call_stack = self.storage().call_stack();
        let caller = call_stack.get(call_stack.len() - 2);
        element_to_key(caller.unwrap_or_revert())
    }
```